### PR TITLE
Animation System Optimizations

### PR DESCRIPTION
**Pull request by hyperlogic**
Small optimizations for the avatar and animation systems including:
* AnimPose operator* and inverse no longer converts scale, rotation and translation into and out of matrix form but instead operates directly on the components.
* more efficient isEqual vec3 check in Rig
* AnimSkeleton::getParentIndex() has better cache locality which avoids frequent cache misses when called in tight loops.
* CubicHermiteSplineWithArcLength used by the spline IK system, was doing redundant calculations and now has less precision to help improve performance.
* Avoid slow dynamic_casts in getAnimInverseKienmaticsNode

None of these changes should have any discernible difference, except for the reduced time spent doing Avatar and Animation tasks.

### CODE REVIEW

File: libraries/animation/src/AnimPose.cpp [hunk 1]
```diff
@@ -40,29 +39,22 @@ glm::vec3 AnimPose::xformPoint(const glm::vec3& rhs) const {
...
 }
 
 AnimPose AnimPose::operator*(const AnimPose& rhs) const {
-    glm::mat4 result;
-    glm_mat4u_mul(*this, rhs, result);
-    return AnimPose(result);
+    float scale = _scale * rhs._scale;
+    glm::quat rot = _rot * rhs._rot;
+    glm::vec3 trans = _trans + (_rot * (_scale * rhs._trans));
+    return AnimPose(scale, rot, trans);
```
Comment (hyperlogic) [inline libraries/animation/src/AnimPose.cpp]: This is the main optimization, operator* was spending a large amount of time converting into and out of matrix form.

File: libraries/animation/src/AnimPose.cpp [hunk 2]
```diff
@@ -40,29 +39,22 @@ glm::vec3 AnimPose::xformPoint(const glm::vec3& rhs) const {
...
 }
 
 AnimPose AnimPose::inverse() const {
-    return AnimPose(glm::inverse(static_cast<glm::mat4>(*this)));
+    float invScale = 1.0f / _scale;
+    glm::quat invRot = glm::inverse(_rot);
+    glm::vec3 invTrans = invScale * (invRot * -_trans);
+    return AnimPose(invScale, invRot, invTrans);
```
Comment (hyperlogic) [inline libraries/animation/src/AnimPose.cpp]: Similarly, here we also avoid converting into and out of matrix form

File: libraries/animation/src/AnimPose.h [hunk 3]
```diff
@@ -50,13 +49,13 @@ class AnimPose {
 
 private:
     friend QDebug operator<<(QDebug debug, const AnimPose& pose);
-    glm::vec3 _scale { 1.0f };
     glm::quat _rot;
     glm::vec3 _trans;
+    float _scale { 1.0f };  // uniform scale only.
```
Comment (hyperlogic) [inline libraries/animation/src/AnimPose.h]: We've removed the vec3 scale and replaced it with a uniform scale factor instead.  This reduces the memory footprint of AnimPoseVecs.  NOTE: non uniform scale was not working and was not supported, so this does not cause us to lose any functionality.

File: libraries/animation/src/AnimSkeleton.h [hunk 4]
```diff
@@ -43,7 +43,10 @@ class AnimSkeleton {
     // get post transform which might include FBX offset transformations
     const AnimPose& getPostRotationPose(int jointIndex) const;
 
-    int getParentIndex(int jointIndex) const;
+    int getParentIndex(int jointIndex) const {
+        return _parentIndices[jointIndex];
+    }
```
Comment (hyperlogic) [inline libraries/animation/src/AnimSkeleton.h]: previously getParentIndex would look into the _joints vector.  The HFMJoint structure is very large, by creating a seperate vector of ONLY indices, this greatly reduces the probability of cache misses when used in tight loops, such as buildAbsoluteRigPoses().

**Comment by hyperlogic** [comment]
https://github.com/highfidelity/hifi/pull/14865

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17095/HighFidelity-Beta-PR14865-5c7e815.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17095/HighFidelity-Beta-PR14865-5c7e815.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17095/HighFidelity-Beta-Interface-PR14865-5c7e815.exe

**Comment by JeffClinton** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17112/HighFidelity-Beta-PR14865-5c7e815.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17112/HighFidelity-Beta-PR14865-5c7e815.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17112/HighFidelity-Beta-Interface-PR14865-5c7e815.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17121/HighFidelity-Beta-PR14865-f860846.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17121/HighFidelity-Beta-PR14865-f860846.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17121/HighFidelity-Beta-Interface-PR14865-f860846.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17121/HighFidelity-Beta-PR14865-f860846.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17121/HighFidelity-Beta-Interface-PR14865-f860846.exe

**Comment by hyperlogic** [comment]
https://highfidelity.manuscript.com/f/cases/20923/Additional-Animation-Optimizations-for-Quest

**Pull request merged**